### ISSUE TITLE
oc_filecache can not be cleaned from old storage entries

### ISSUE THREAD

**Issue by fra-orolo** [nextcloud/server#27365]
I wanted to upgrade from NC 20 to 21, but the snap upgrade on Ubuntu did not work.
Somewhere in the logs I found, that a select count(*) form oc_filecache ran into a timeout.

### Steps to reproduce
1. create a external SFTP storage with a lot of files
2. scan the storage, look for the storage id in oc_storages
3. remove the storage, occ files:cleanup
4. search for old entries in oc_filecache:
select * from oc_filecache where storage not in (select select numeric_id from oc_storages) limit 100;

### Expected behaviour
Only entries from existing storages should be found in oc_filecache

### Actual behaviour
There are millions of entries for the deleted storage left.
This prevents me from a full row count
select count(fileid) from oc_filecache.
The statement did not answer while running over night.
 
### Server configuration

**Operating system:**
Ubuntu 18, nextcloud snap image
**Web server:**
Apache
**Database:**
MySql Innodb

**PHP version:**

**Nextcloud version:** (see Nextcloud admin page)
20.0.9.1

**Updated from an older Nextcloud/ownCloud or fresh install:**
updated several times 

**Where did you install Nextcloud from:**

**Signing status:**
<details>
<summary>Signing status</summary>

```
Login as admin user into your Nextcloud and access 
http://example.com/index.php/settings/integrity/failed 
paste the results here.
```
</details>

**List of activated apps:**
<details>
<summary>App list</summary>

```
If you have access to your command line run e.g.:
sudo -u www-data php occ app:list
from within your Nextcloud installation folder
```
</details>

**Nextcloud configuration:**
<details>
<summary>Config report</summary>

```
If you have access to your command line run e.g.:
sudo -u www-data php occ config:list system
from within your Nextcloud installation folder

or 

Insert your config.php content here. 
Make sure to remove all sensitive content such as passwords. (e.g. database password, passwordsalt, secret, smtp password, …)
```
</details>

**Are you using external storage, if yes which one:** local/smb/sftp/...

**Are you using encryption:** yes/no

**Are you using an external user-backend, if yes which one:** LDAP/ActiveDirectory/Webdav/...

#### LDAP configuration (delete this part if not used)
<details>
<summary>LDAP config</summary>

```
With access to your command line run e.g.:
sudo -u www-data php occ ldap:show-config
from within your Nextcloud installation folder

Without access to your command line download the data/owncloud.db to your local
computer or access your SQL server remotely and run the select query:
SELECT * FROM `oc_appconfig` WHERE `appid` = 'user_ldap';


Eventually replace sensitive data as the name/IP-address of your LDAP server or groups.
```
</details>

### Client configuration
**Browser:**

**Operating system:**

#### Web server error log
<details>
<summary>Web server error log</summary>

```
Insert your webserver log here
```
</details>

#### Nextcloud log (data/nextcloud.log)
<details>
<summary>Nextcloud log</summary>

```
Insert your Nextcloud log here
```
</details>

#### Browser log
<details>
<summary>Browser log</summary>

```
Insert your browser log here, this could for example include:

a) The javascript console log
b) The network log
c) ...
```
</details>

**Comment by szaimen** [comment]
I suppose this still happens with NC21.0.3?

**Comment by fra-orolo** [comment]
Ahh, sorry.
What I found for my system was finally, that one of the external file systems contained symlink-loops (on a linux system).
Nextcloud tried to scan this and ended up with a huge amount of entries in the cache.
I ended up with the following:
* dropping that external file system
* copying all entries from the filecache table, that belonged to something else
* truncating (not delete) the original table
* and copying the good entries back from the other table.
